### ISSUE TITLE
namespaceSelector:{} allows connections from external Load Balancer

### ISSUE THREAD

**Issue by ahmetb** [projectcalico/k8s-policy#113]
## Expected Behavior

If I apply a network policy to a pod exposed to internet via service.type=LoadBalancer with an ingress rule like:

    ingress:
    - from:
        - namespaceSelector: {}

I expect it to block traffic from the external sources, such as the external LB. However, I expect it to continue allowing traffic from pod in all namespaces.

## Current Behavior

Pods exposed via Service.type=LoadBalancer still accept connections from external sources when a policy that allows only `namespaceSelector: {}` is applied.

## Possible Solution

Unsure.

## Steps to Reproduce (for bugs)
1. Create a GKE alpha 1.7.2 cluster with network policy enabled.
2. `kubectl run --image=nginx`
3. `kubectl expose deployment web --type=LoadBalancer`
4. `curl IP` developer workstation (get IP from `kubectl get svc`), note it works.
5. Apply manifest:

```yaml
apiVersion: extensions/v1beta1
kind: NetworkPolicy
metadata:
  name: web-allow-all
spec:
  podSelector:
    matchLabels:
      run: web
  ingress:
  - from:
      - namespaceSelector: {}
```

6. Note that `curl IP` from developer workstation still works, but it should not.

## Your Environment

* Kubernetes version (`kubectl version`): 1.7.2
* Manifest file used to deploy k8s policy: https://gist.github.com/ahmetb/096f035750b0721768cad0c6dc3242f0
* Operating System and version: Container-optimized OS from Google, Kernel 4.4.52+, Docker 1.11.2

cc: @dnardo @thockin

**Comment by caseydavenport** [comment]
Ok, @gunjan5 and I have reproduced this and understand the cause.

Essentially, it's a bug in the NetworkPolicy conversion logic in the Kubernetes datastore driver which handles the `namespaceSelector`, which is not properly converting the "select all namespaces" case.

@gunjan5 should have a fix up soon.

**Comment by caseydavenport** [comment]
This has been fixed in Calico v2.4.1: https://github.com/projectcalico/calico/releases/tag/v2.4.1
